argoverse/utils/make_att_files.py

```diff
@@ -0,0 +1,365 @@
+                int(bbox_2d[edge_2d[ii][0], 1] * image_scale + image_size / 2),
+                int(bbox_2d[edge_2d[ii][0], 0] * image_scale + image_size / 2),
+            )
+            p2 = (
+                int(bbox_2d[edge_2d[ii][1], 1] * image_scale + image_size / 2),
+                int(bbox_2d[edge_2d[ii][1], 0] * image_scale + image_size / 2),
+            )
+            cv2.line(img, p1, p2, color=color)
+
+    kernel = np.ones((5, 5), np.float)
+    img = cv2.dilate(img, kernel, iterations=1)
+    cv2.putText(
+        img,
+        "%s_%s_%d" % (dataset_name, log_id, lidar_timestamp),
+        (100, image_size - 100),
+        font,
+        fontScale,
+        fontColor,
+        lineType,
+    )
+
+    offset = 0
+    for key in dict_color.keys():
+        cv2.putText(img, key, (100 + offset, image_size - 50), font, fontScale, dict_color[key], lineType)
+        offset += 150
+
+    print("Saving img: ", path_imgs)
+    cv2.imwrite(os.path.join(path_imgs, "%s_%s_%d.jpg" % (dataset_name, log_id, lidar_timestamp)), img * 255)
+
+
+def bspline_1d(x: np.array, y: np.array, s: float = 20.0, k: int = 3) -> np.array:
+    """
+    Do Bspline smoothing
+    """
+
+    if len(x) <= k:
+        return y
+
+    tck = interpolate.splrep(x, y[x], s=s, k=k)
+
+    return interpolate.splev(np.arange(y.shape[0]), tck)
+
+
+def derivative(x: np.array) -> np.array:  # x: N-length np array
+    """
+    Compute derivative for velocity and acceleration 
+    """
+    x_tensor = torch.Tensor(x).unsqueeze(0).unsqueeze(0)
+    x_padded = torch.cat((x_tensor, (x_tensor[:, :, -1] - x_tensor[:, :, -2] + x_tensor[:, :, -1]).unsqueeze(0)), dim=2)
+    filters = torch.Tensor([-1, 1]).unsqueeze(0).unsqueeze(0)
+
+    return F.conv1d(x_padded, filters)[0, 0].numpy()
+
+
+def compute_v_a(traj: np.array) -> Tuple[np.array, np.array]:  # traj:Nx3
+    """
+    Compute velocity and acceleration
+    """
+
+    ind_valid = np.nonzero(1 - np.isnan(traj[:, 0]))[0]
+    dx = bspline_1d(ind_valid, traj[:, 0])
+    dy = bspline_1d(ind_valid, traj[:, 1])
+    dz = bspline_1d(ind_valid, traj[:, 2])
+
+    if len(ind_valid) > 0:
+        x_max = dx[ind_valid].max()
+        x_min = dx[ind_valid].min()
+        y_max = dy[ind_valid].max()
+        y_min = dy[ind_valid].min()
+
+        if np.sqrt((x_max - x_min) ** 2 + (y_max - y_min) ** 2) < 10:
+            return np.zeros_like(traj), np.zeros_like(traj)
+
+    vx = derivative(dx) * 10  # lidar freq 10 hz
+    vy = derivative(dy) * 10
+    vz = derivative(dz) * 10
+
+    ax = derivative(vx) * 10
+    ay = derivative(vy) * 10
+    az = derivative(vz) * 10
+
+    v = np.concatenate((vx[:, np.newaxis], vy[:, np.newaxis], vz[:, np.newaxis]), axis=1)
+    a = np.concatenate((ax[:, np.newaxis], ay[:, np.newaxis], az[:, np.newaxis]), axis=1)
+
+    v[np.isnan(traj[:, 0])] = np.nan
+    a[np.isnan(traj[:, 0])] = np.nan
+    return v, a
+
+
+if __name__ == "__main__":
+    # set root_dir to the correct path to your dataset folder
```

Comment:
maybe we could make this long function a separate function named `make_att_files()`?